Set BuildKit version and revision at build time

crazy-max opened the pull request:
follow-up https://github.com/moby/moby/pull/44686#discussion_r1094416821

**- What I did**

We should set the BuildKit version and revision at build time so we make sure we are aligned with the expected vendoring.

Also needs to remove the dependency on git in `with-go-mod.sh` script as we ignore `.git` folder in `.dockerignore`, otherwise it fails. And needs to make bind mount writable for the `go.mod` file as well.

**- How I did it**

Use go mod tooling and ldflags to set BuildKit version at build time.

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

crazy-max commented on dockerversion/version_lib.go:
```diff
@@ -9,4 +9,5 @@ var (
 	PlatformName          = ""
 	ProductName           = ""
 	DefaultProductLicense = ""
+	BuildKitVersion       = "v0.11.2+unknown"
```
@thaJeztah I think it would be handy to also add this to `docker info` in follow-up. WDYT?

neersighted replied:
I feel like we should avoid adding more versions to `docker info` -- the Go psuedo-versions/SHAs presented there are quite unhelpful/confusing to users as they are presented as e.g. 'containerd version: SHA'

I would suggest instead that you add reporting of the BuildKit version to `docker version`, which is designed to be modular and doesn't require a new API to add this information.

crazy-max committed:
Set BuildKit version and revision at build time

We should set the BuildKit version and revision at build time so
we make sure we are aligned with the expected vendoring.

Also needs to remove the dependency on git in with-go-mod script as we
ignore .git folder in dockerignore, otherwise it fails. And needs to
make bind mount writable for the go.mod file as well.

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah commented on builder/builder-next/worker/worker.go:
```diff
@@ -50,7 +51,8 @@ import (
 
 func init() {
-	version.Version = "v0.11.0-rc3"
+	version.Version = dockerversion.BuildKitVersion
```
Wondering; instead of having this as a package level variable (and adding dockerversion as a dependency here), would it make sense to store this on the Worker itself, and pass it as an argument/option in the Worker constructor?

tianon commented on builder/builder-next/worker/worker.go:
```diff
@@ -49,8 +49,16 @@ import (
 	"golang.org/x/sync/semaphore"
 )
 
+var (
+	// buildkitVersion holds the version number. Filled in at linking time.
+	buildkitVersion = "v0.0.0+unknown"
+	// buildkitRevision holds the VCS revision. Filled in at linking time.
+	buildkitRevision = ""
+)
```
It's really a shame we can't get this directly from `go.mod` (even via something like https://pkg.go.dev/runtime/debug#ReadBuildInfo, but it could be more direct if it were a first-order language feature).

Maybe we could do something like abusing `go:embed` to embed the `go.mod` itself and re-parse it? :confused:

(Overall, I think we should trend more towards a raw `go build` being "enough" for the most part, so anything that requires movement in the other direction should be weighed carefully IMO)

Would love to hear more maintainer opinions :bow:

crazy-max replied:
> (Overall, I think we should trend more towards a raw `go build` being "enough" for the most part, so anything that requires movement in the other direction should be weighed carefully IMO)

Yeah I also don't really like having this logic in the build pipeline but this information is still useful for analytics and debugging purpose when someone reports an issue imo. If only go tooling could provide git information for a dependency that would be handy.

jedevc replied:
Is there not a way to get this from `ReadBuildInfo`?

I would have thought `BuildInfo.Deps[].Version` would have the information we want? I know we can't actually use this easily in BuildKit itself, since there's no way to actually get the version for the main module (it's left empty in `BuildInfo.Main`), but I think we should be able to use it here fairly easily?
